### PR TITLE
Inject basketfactory for payment controller

### LINKED ISSUE #624: On payment validation - The "sonata.basket.factory" service or alias has been removed
## Subject

The "sonata.basket.factory" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.

## Steps to reproduce
after payment validation get error message

## Expected results
Payment accepted

### PR DESCRIPTION
**Pull request by salamichel**
## Subject
The "sonata.basket.factory" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.

Fixes #624 

## Changelog

```markdown
### Fixed
- crash on payment validation
```

### CODE REVIEW

File: src/PaymentBundle/Controller/PaymentController.php [hunk 1]
```diff
@@ -27,6 +27,16 @@
 class PaymentController extends Controller
 {
     /**
+     * @var BasketFactoryInterface
+     */
+    private $basketFactory;
+
+    public function __construct(BasketFactoryInterface $basketFactory)
```
Comment (core23) [inline src/PaymentBundle/Controller/PaymentController.php]: This is a BC break on the stable branch. Please make it nullable

File: src/PaymentBundle/Controller/PaymentController.php [hunk 2]
```diff
@@ -26,6 +26,16 @@
...
 class PaymentController extends Controller
 {
+    /**
+     * @var BasketFactoryInterface
+     */
+    private $basketFactory;
+
+    public function __construct(BasketFactoryInterface $basketFactory = null)
+    {
+        $this->basketFactory = $basketFactory;
```
Comment (greg0ire) [inline src/PaymentBundle/Controller/PaymentController.php]: I think we should also trigger a deprecation error here if `$basketFactory` is `null`

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\PaymentBundle\Tests\Controller;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\Component\Basket\BasketFactoryInterface;
+use Sonata\PaymentBundle\Controller\PaymentController;
+
+class PaymentControllerTest extends TestCase
```
Comment (core23) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
final class PaymentControllerTest extends TestCase
```

File: src/PaymentBundle/Controller/PaymentController.php (hunk 2 above)
Comment (salamichel) [reply src/PaymentBundle/Controller/PaymentController.php]: you mean use something @trigger_error what kind of message should i put?

**Comment by salamichel** [comment]
@core23 @greg0ire Hi guys I do my arm on tools like continous-integration / coveralls code etc thank you for your support. i try to some code coveralls on paymentController for starting how to test `return $this->get('sonata.basket.factory');` i get assert error because get return null (service not found i persume).

File: src/PaymentBundle/Controller/PaymentController.php (hunk 2 above)
Comment (greg0ire) [reply src/PaymentBundle/Controller/PaymentController.php]: Yes, there are tons of examples you can find in Sonata if you grep for `E_USER_DEPRECATED`. In your case,
```
@trigger_error(sprintf(
    'Not providing a %s instance to %s is deprecated since sonata-project/ecommerce 3.x. Providing it will be mandatory in 4.0',
    BasketFactoryInterface::class,
    E_USER_DEPRECATED
));
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        $this->assertSame($basketFactoryInterface, $method->invoke($paymentController));
+    }
+
+    public function createPaymentController($basketFactory = null)
+    {
+        if (null === $basketFactory) {
+            $basketFactory = $this->createMock(BasketFactoryInterface::class);
+        }
+
+        return new PaymentController($basketFactory);
```
Comment (greg0ire) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]: The tests are failing because you need to pass all dependencies or mark this test as legacy (with `@group legacy`)

File: src/PaymentBundle/Controller/PaymentController.php [hunk 5]
```diff
@@ -161,14 +211,14 @@ protected function getBasketFactory()
      */
     protected function getBasket()
     {
-        return $this->get('sonata.basket');
+        return $this->basket;
```
Comment (pierredup) [inline src/PaymentBundle/Controller/PaymentController.php]: `$this->basket` may be `null` here, so you might want to add the same check as in `getBasketFactory` (same below in `getPaymentHandler`)

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        $basketFactoryInterface = $this->createMock(BasketFactoryInterface::class);
+
+        $paymentController = $this->createPaymentController($basketFactoryInterface);
+
+        $this->assertSame($basketFactoryInterface, $method->invoke($paymentController));
+    }
+
+    public function createPaymentController($basketFactory = null)
+    {
+        if (null === $basketFactory) {
```
Comment (pierredup) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]: `$basketFactory` won't ever be `null` here, since you are explicitly passing a mock object through

File: src/PaymentBundle/Controller/PaymentController.php (hunk 5 above)
Comment (salamichel) [reply src/PaymentBundle/Controller/PaymentController.php]: You're so right too quick last modification

File: tests/PaymentBundle/Controller/PaymentControllerTest.php (hunk 6 above)
Comment (salamichel) [reply tests/PaymentBundle/Controller/PaymentControllerTest.php]: Yes you right too, but in fact i tried to do another test by calling createPaymentController(null) for testing this part https://github.com/sonata-project/ecommerce/pull/625/files#diff-0c04180e3dc418019a69461490a25e6cR206

but i give up don't know how to test this

Comment (pierredup) [reply tests/PaymentBundle/Controller/PaymentControllerTest.php]: You will need to create a mock `ContainerInterface` and set it on the controller, in order to test getting the service from the container. Something like this

```php
$container = $this->createMock(ContainerInterface::class);
$container->expects($this->once())
    ->method('get')
    ->with('sonata.basket.factory')
    ->willReturn( $basketFactoryInterface);

$paymentController->setContainer($container);
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        $methodGetBasket->setAccessible(true);
+
+        $methodGetPaymentHandler = $classPaymentController->getMethod('getPaymentHandler');
+        $methodGetPaymentHandler->setAccessible(true);
+
+        $basketFactoryInterface = $this->createMock(BasketFactoryInterface::class);
+        $PaymentHandlerInterface = $this->createMock(PaymentHandlerInterface::class);
+        $Basket = $this->createMock(Basket::class);
+
+        $paymentController = $this->createPaymentController();
```
Comment (pierredup) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
        $paymentController = new PaymentController($basketFactory, $paymentHandler, $basket)
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        $methodGetBasketFactory->setAccessible(true);
+
+        $methodGetBasket = $classPaymentController->getMethod('getBasket');
+        $methodGetBasket->setAccessible(true);
+
+        $methodGetPaymentHandler = $classPaymentController->getMethod('getPaymentHandler');
+        $methodGetPaymentHandler->setAccessible(true);
+
+        $basketFactoryInterface = $this->createMock(BasketFactoryInterface::class);
+        $PaymentHandlerInterface = $this->createMock(PaymentHandlerInterface::class);
```
Comment (pierredup) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
        $paymentHandlerInterface = $this->createMock(PaymentHandlerInterface::class);
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        $Basket = $this->createMock(Basket::class);
```
Comment (pierredup) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
        $basket = $this->createMock(Basket::class);
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        $methodGetBasketFactory->setAccessible(true);
+
+        $methodGetBasket = $classPaymentController->getMethod('getBasket');
+        $methodGetBasket->setAccessible(true);
+
+        $methodGetPaymentHandler = $classPaymentController->getMethod('getPaymentHandler');
+        $methodGetPaymentHandler->setAccessible(true);
+
+        $basketFactoryInterface = $this->createMock(BasketFactoryInterface::class);
+        $Basket = $this->createMock(Basket::class);
```
Comment (pierredup) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
        $basket = $this->createMock(Basket::class);
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        $PaymentHandlerInterface = $this->createMock(PaymentHandlerInterface::class);
```
Comment (pierredup) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
        $paymentHandlerInterface = $this->createMock(PaymentHandlerInterface::class);
```

**Comment by salamichel** [comment]
@core23 Hi are you still waiting something from me ?

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        $paymentController = new PaymentController();
+
+        $paymentController->setContainer($container);
+
+        $this->assertSame($basketFactoryInterface, $methodGetBasketFactory->invoke($paymentController));
+        $this->assertSame($basket, $methodGetBasket->invoke($paymentController));
+        $this->assertSame($paymentHandlerInterface, $methodGetPaymentHandler->invoke($paymentController));
+    }
+
+    public function createPaymentController(BasketFactoryInterface $basketFactory = null, PaymentHandlerInterface $paymentHandler = null, Basket $basket = null)
```
Comment (core23) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]: Why is this public?

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+        $this->assertSame($basketFactoryInterface, $methodGetBasketFactory->invoke($paymentController));
+        $this->assertSame($paymentHandlerInterface, $methodGetPaymentHandler->invoke($paymentController));
+        $this->assertSame($basket, $methodGetBasket->invoke($paymentController));
+    }
+
+    /**
+     * @group legacy
+     */
+    public function testInstanceNullBasketInterfacePaymentAction()
```
Comment (greg0ire) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
    public function testInstanceNullBasketInterfacePaymentAction(): void
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        $paymentController = new PaymentController();
+
+        $paymentController->setContainer($container);
+
+        $this->assertSame($basketFactoryInterface, $methodGetBasketFactory->invoke($paymentController));
+        $this->assertSame($basket, $methodGetBasket->invoke($paymentController));
+        $this->assertSame($paymentHandlerInterface, $methodGetPaymentHandler->invoke($paymentController));
+    }
+
+    public function testSendBankAction()
```
Comment (greg0ire) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
    public function testSendBankAction(): void
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+
+        $paymentController = $this->createPaymentController(null, $paymentHandlerInterface, $basket);
+
+        $this->assertInstanceOf(Response::class, $paymentController->sendbankAction($request));
+    }
+
+    /**
+     * @return \Sonata\Component\Order\OrderInterface
+     */
+    protected function getOrder()
```
Comment (greg0ire) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]: Please remove the phpdoc, add a use statement and  a return type declaration:

```suggestion
    private function getOrder(): OrderInterface
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+        $order->setReference('my-reference');
+
+        return $order;
+    }
+
+    /**
+     * @return \Sonata\Component\Payment\Debug\DebugPayment
+     */
+    protected function getDebugPayment()
```
Comment (greg0ire) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
    private function getDebugPayment(): DebugPayment
```

File: tests/PaymentBundle/Controller/PaymentControllerTest.php [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        $client = $this->createMock(ClientInterface::class);
+
+        $browser = new Browser($client);
+
+        $payment = new DebugPayment($router, $browser);
+
+        return $payment;
+    }
+
+    protected function createPaymentController(BasketFactoryInterface $basketFactory = null, PaymentHandlerInterface $paymentHandler = null, Basket $basket = null)
```
Comment (greg0ire) [inline tests/PaymentBundle/Controller/PaymentControllerTest.php]:
```suggestion
    private function createPaymentController(
        ?BasketFactoryInterface $basketFactory = null, 
        ?PaymentHandlerInterface $paymentHandler = null,
        ?Basket $basket = null
    ): PaymentController
```

**Comment by greg0ire** [comment]
Please remove

```

Update tests/PaymentBundle/Controller/PaymentControllerTest.php

commit suggestion
```

From your commit message

**Comment by greg0ire** [comment]
Please write a changelog directed at end users.
More on http://keepachangelog.com

**Comment by salamichel** [comment]
@greg0ire i have modified as this

### Changed
- update PaymentController inject 'sonata.payment.handler' / 'sonata.basket' / 'sonata.basket.factory' services
- add some paymentController tests cases

i really don't know what to explain, i just want to make it work and correct #624

**Comment by greg0ire** [comment]
Ok I'm going to change the changelog for you, watch me

**Comment by greg0ire** [comment]
Thanks @salamichel !
